Information about SQLite limitations

usmangt opened the pull request:
I am creating this PR to add more content about the SQLite (the default database for Grafana) limitations. SQLite is fine for small environments but not suitable when monitoring many Dashboards and sending lots of alerts frequently.

For e.g. this issue is just from this week and then got followed up by 3-4 more similar issues (all related to SQLite and are linked with this issue)

Closes #64664 

By adding more information, will have less frequency of such issues reported by users.

Also, once this PR get submitted, will also create a new PR about the possible workaround to not get the error message "database locked" while using SQLite.

<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

7. If your PR content should be added to the What's New document for the next major or minor release, add the **add to what's new** label to your PR. Note that you should add this label to the main PR that introduces the feature; do not add this label to smaller PRs for the feature.

-->

**What is this feature?**

[Add a brief description of what the feature or update does.]

**Why do we need this feature?**

[Add a description of the problem the feature is trying to solve.]

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #

**Special notes for your reviewer**:

usmangt committed:
Update _index.md

I am creating this PR to add more content about the SQLite (the default database for Grafana) limitations. SQLite is fine for small environments but not suitable when monitoring many Dashboards and sending lots of alerts frequently.

For e.g. this issue is just from this week and then got followed up by 3-4 more similar issues (all related to SQLite and are linked with this issue)

Closes #64664 

By adding more information, will have less frequency of such issues reported by users.

Also, once this PR get submitted, will also create a new PR about the possible workaround to not get the error message "database locked" while using SQLite.

mbrancato commented:
@usmangt I think it would be good to describe I/O requirements. Realistically, Grafana needs very little storage.  In cloud deployments such as on Kubernetes, small PVCs = small IOPS / disk bandwidth. It does seem that without WAL mode, Grafana is not compatible with slow disks, even in moderate alerting usage. 

I did go through my history, the SQLite limitations didn't happen till after Grafana 9.x where we were forced onto the new alerting system.

It appears as if under the default operation, the SQLite database allows writers to block readers, and this appears to return an error. I think that Grafana should consider enabling WAL mode by default, given that without it, there is no MVCC implementation - writers are going to block readers and general concurrency issues will occur. It's a 12+ year old feature, so I think its probably fairly stable.
https://www.sqlite.org/wal.html

See my other issue on the wrong / poorly descriptive error type being returned: #64692

usmangt commented:
Hi @mbrancato 

Thanks for the feedback.

My main objective of this PR is to provide some more information when using SQLite so that users who not do know in advance which backend Database they want to choose accordingly before installing Grafana and scaling it out and started to run into issues (if their systems are in production environments). 

For e g.especially in [HA environments](https://grafana.com/docs/grafana/latest/setup-grafana/set-up-for-high-availability/) where you have to switch to either MySQL or PostgreSQL. (I will add this part in this PR so as to make it more visible).


Another example is that if someone is using containerized environments such as Kubernetes etc. and get such an error (or wants HA) then have to go all the way to migrate their database from SQLite to either MySQL or PostgreSQL (which is off-course a downtime for them and need to tackle it before proceeding forward).


> I think that Grafana should consider enabling WAL mode by default

Yes, that could be considered and I also wanted to anyway write this part in our troubleshooting guide. But I will create a request for our Engineering team and they will review it. :+1: 

> See my other issue on the wrong / poorly descriptive error type being returned: https://github.com/grafana/grafana/issues/64692

I am already working on that one :slightly_smiling_face: and saw your new reply (will follow up there)

mbrancato commented:
@usmangt I think that regardless of how many alerts are setup, this is an error users could see. I increased the IOPS on my disk by 5x and I still saw this error (only once so far). If there is any concurrency in Grafana (its using Go, so likely yes here), you're always going to be working in a race condition between multiple access attempts to the database.

What you're saying here is it is just "less likely" - the race condition still exists as Grafana uses concurrency but the DB has no concurrency control mechanism. Instead of documenting this, I think its important to understand the underlying issue is not documentation, but implementation. That might require that Grafana itself use concurrency control mechanisms like a shared semaphore or mutex to control access to the DB to avoid concurrency when there is no MVCC.

https://pkg.go.dev/sync#Mutex

chri2547 reviewed (changes requested):
Thanks for the contribution! Please take a look at my suggestion and see how that works for you. When you are ready for me to approve, please ask me to re-review.

usmangt committed:
Update docs/sources/setup-grafana/installation/_index.md

Thank you. Also, can you please add this link to the High Availablity text as hyperlink.

Co-authored-by: Christopher Moyer <35463610+chri2547@users.noreply.github.com>

chri2547 committed:
Update docs/sources/setup-grafana/installation/_index.md

chri2547 commented on docs/sources/setup-grafana/installation/_index.md:
```diff
@@ -54,6 +54,8 @@ Grafana supports the following databases:
 
 By default, Grafana installs with and uses SQLite, which is an embedded database stored in the Grafana installation location.
 
+> **Note:** SQLite works well if your environment is small, but is not recommended when your environment starts growing. For more information about the limitations of SQLite, refer to [Appropriate Uses For SQLite](https://www.sqlite.org/whentouse.html). If you want high availability, you must use either a MySQL or PostgreSQL database. For information about how to define the database configuration ...
```
```suggestion
> **Note:** SQLite works well if your environment is small, but is not recommended when your environment starts growing. For more information about the limitations of SQLite, refer to [Appropriate Uses For SQLite](https://www.sqlite.org/whentouse.html). If you want high availability, you must use a MySQL or PostgreSQL database. For information about how to define the database configuration parameters inside the `grafana.ini` file, refer to [[database]](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#database).
```

chri2547 committed:
Update docs/sources/setup-grafana/installation/_index.md

chri2547 reviewed (approved):
Looks great! Thank you for the contribution!

GeorgeMaged commented:
Hello there, I am using Grafana 9.3.6 with the default sqlite3 DB, And after creating alarms that are monitoring a large number of network interfaces if they exceed the threshold or not, Since then Grafana started crashing and I found many logs having the error " database is locked" . I tried when I read here in the community enabling the WAL but It didn't make such a difference. Do I need to consider migrating into mysql or postgres?
